core/drivers/atmel_wdt.c

```diff
@@ -0,0 +1,222 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2022 Microchip
+ */
+
+#include <assert.h>
+#include <drivers/wdt.h>
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <io.h>
+#include <matrix.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <sama5d2.h>
+#include <tee_api_types.h>
+
+#define WDT_CR			0x0
+#define WDT_CR_KEY		SHIFT_U32(0xA5, 24)
+#define WDT_CR_WDRSTT		BIT(0)
+
+#define WDT_MR			0x4
+#define WDT_MR_WDV		GENMASK_32(11, 0)
+#define WDT_MR_WDV_SET(val)	((val) & WDT_MR_WDV)
+#define WDT_MR_WDFIEN		BIT(12)
+#define WDT_MR_WDRSTEN		BIT(13)
+#define WDT_MR_WDDIS		BIT(15)
+#define WDT_MR_WDD_SHIFT	16
+#define WDT_MR_WDD_MASK		GENMASK_32(11, 0)
+#define WDT_MR_WDD		SHIFT_U32(WDT_MR_WDD_MASK, WDT_MR_WDD_SHIFT)
+#define WDT_MR_WDD_SET(val) \
+			SHIFT_U32(((val) & WDT_MR_WDD_MASK), WDT_MR_WDD_SHIFT)
+#define WDT_MR_WDDBGHLT		BIT(28)
+#define WDT_MR_WDIDLEHLT	BIT(29)
+
+#define WDT_SR			0x8
+#define WDT_SR_DUNF		BIT(0)
+#define WDT_SR_DERR		BIT(1)
+
+/*
+ * The watchdog is clocked by a 32768Hz clock/128 and the counter is on
+ * 12 bits.
+ */
+#define SLOW_CLOCK_FREQ		(32768)
+#define WDT_CLOCK_FREQ		(SLOW_CLOCK_FREQ / 128)
+#define WDT_MIN_TIMEOUT		1
+#define WDT_MAX_TIMEOUT		(BIT(12) / WDT_CLOCK_FREQ)
+
+#define WDT_DEFAULT_TIMEOUT	WDT_MAX_TIMEOUT
+
+/*
+ * We must wait at least 3 clocks period before accessing registers MR and CR.
+ * Ensure that we see at least 4 edges
+ */
+#define WDT_REG_ACCESS_UDELAY	(1000000ULL / SLOW_CLOCK_FREQ * 4)
+
+#define SEC_TO_WDT(sec)		(((sec) * WDT_CLOCK_FREQ) - 1)
+
+#define WDT_ENABLED(mr)		(!((mr) & WDT_MR_WDDIS))
+
+struct atmel_wdt {
+	struct wdt_chip chip;
+	vaddr_t base;
+	unsigned long rate;
+	uint32_t mr;
+};
+
+static void atmel_wdt_write_sleep(struct atmel_wdt *wdt, uint32_t reg,
+				  uint32_t val)
+{
+	udelay(WDT_REG_ACCESS_UDELAY);
```

Comment:
You should perhaps call `wait_cycles()` instead with the number of cycles you'd like to wait.

Reply:
Hum, wouldn't this be breaking the fact that the driver can work on multiple platforms ? That why I added the udelay layer that uses the platform frequency.

Reply:
Sorry, my mistake.